// ts-postgres/dist/commonjs/defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Defaults = void 0;
const node_os_1 = require("node:os");
function secToMsec(value) {
    if (typeof value === 'number' && !isNaN(value)) {
        return value * 1000;
    }
}
class Defaults {
    constructor(env, host = env.PGHOST || 'localhost', port = parseInt(env.PGPORT, 10) || 5432, user = env.PGUSER || (0, node_os_1.userInfo)().username, database = env.PGDATABASE, password = env.PGPASSWORD, preparedStatementPrefix = 'tsp_', sslMode = env.PGSSLMODE, connectionTimeout = secToMsec(parseInt(env.PGCONNECT_TIMEOUT, 10)), clientEncoding = env.PGCLIENTENCODING) {
        this.host = host;
        this.port = port;
        this.user = user;
        this.database = database;
        this.password = password;
        this.preparedStatementPrefix = preparedStatementPrefix;
        this.sslMode = sslMode;
        this.connectionTimeout = connectionTimeout;
        this.clientEncoding = clientEncoding;
    }
}
exports.Defaults = Defaults;
